library/modules/EventManager.cpp

```diff
@@ -491,31 +497,39 @@ static void manageJobStartedEvent(color_ostream& out) {
     // iterate event handler callbacks
     multimap<Plugin*, EventHandler> copy(handlers[EventType::JOB_STARTED].begin(), handlers[EventType::JOB_STARTED].end());
 
-    unordered_set<int32_t> newStartedJobs;
-
-    for (df::job_list_link* link = &df::global::world->jobs.list; link->next != nullptr; link = link->next) {
-        df::job* job = link->next->item;
-        if (!job || !Job::getWorker(job))
-            continue;
-
-        int32_t j_id = job->id;
-        newStartedJobs.emplace(j_id);
-        if (!startedJobs.count(j_id)) {
-            for (auto &[_,handle] : copy) {
-                DEBUG(log,out).print("calling handler for job started event\n");
-                run_handler(out, EventType::JOB_STARTED, handle, job);
+    std::vector<int32_t> newStartedJobs;
+    newStartedJobs.reserve(startedJobs.size());
+
+    for (const auto jobPtr : df::global::world->jobs.list) {
+        // posting_index of -1 implies a worker has been assigned to a new job.
+        if (jobPtr->posting_index == -1) {
+            auto jobId = jobPtr->id;
+            newStartedJobs.push_back(jobId);
```

Comment:
are jobs in the list guaranteed to be ordered by id?

my testing shows that they appear to be ordered by id, even with repeat jobs, but I don't know if this is a guarantee. @ab9rf do you know?

Reply:
the game doesn't intentionally sort the list but the manner in which jobs are created by the game does (currently) guarantees that they will be in order

Reply:
that implies he doesn't perform any binary searches on the job list.  We can always sort the vector ourselves, if we want to be sure.

Reply:
> that implies he doesn't perform any binary searches on the job list. We can always sort the vector ourselves, if we want to be sure.

if he does i've never found evidence of it. i've also never seen any code for inserting elements anywhere other than the end of the list

Reply:
I'm not aware of any use case where he'd need to binary search the list. jobs are normally referenced as `df::job *`, not by id